### ISSUE TITLE
Incorrect errors for fstrings + generators

### ISSUE THREAD

**Issue by ns-cweber** [python/mypy#3385]
mypy version: 0.511
python version: 3.6.1

Given this valid program:

``` py
from typing import Iterator

def stringify(it: Iterator[int]) -> Iterator[str]:
    return (f"{x}" for x in it)

def stringify2(it: Iterator[int]) -> Iterator[str]:
    for x in it:
        yield f"{x}"

def stringify3(it: Iterator[int]) -> Iterator[str]:
    return (str(x) for x in it)

def stringify4(it: Iterator[int]) -> Iterator[str]:
    for x in it:
        yield str(x)
```

I get errors for the first two variations:

```
foo.py:4: error: Generator has incompatible item type "int"
foo.py:8: error: Incompatible types in yield (actual type "int", expected type "str")
```

It seems that mypy doesn't recognize that fstrings are produce a `str`. Interestingly, the yield error didn't show up in my old mypy version 0.501.

**Comment by ns-cweber** [comment]
Strangely, this is fine:

```py
def stringify5(it: Iterator[Tuple[int, int]]) -> Iterator[str]:
    return (f"({x}, {y}" for x, y in it)
```

**Comment by ns-cweber** [comment]
It looks like it doesn't have anything to do with the unpacking; it seems to pertain to the handling of `ast.FormattedValue`. Mypy handles this correctly:

```py
def stringify(it: Iterator[int]) -> Iterator[str]:
    return (f"asdf {x}" for x in it)
```

According to [GreenTreeSnakes' FormattedValue docs][0]:

> If the string contains a single formatting field and nothing else the node can be isolated otherwise it appears in JoinedStr.

In this most recent example, using an fstring with tokens besides the formatted value produces a JoinedStr token. Maybe mypy handles JoinedStr correctly but not FormattedValue?

**Comment by JelleZijlstra** [comment]
Yes, it looks like the bug is that mypy interprets `f"{x}"` as having the type of `x`.
